File: supermarket-pricing/python/tests/checkout_builder.py
from __future__ import annotations

from dataclasses import dataclass

from supermarket_pricing import Checkout, ComboDeal, Money, Product, Weight


@dataclass
class _Scanned:
    product: Product
    count: int


@dataclass
class _Weighed:
    product: Product
    weight: Weight


class CheckoutBuilder:
    def __init__(self) -> None:
        self._scanned: list[_Scanned] = []
        self._weighed: list[_Weighed] = []
        self._combo_deals: list[ComboDeal] = []

    def with_scanned(self, product: Product, count: int = 1) -> CheckoutBuilder:
        self._scanned.append(_Scanned(product, count))
        return self

    def with_weighed(self, product: Product, kg: float | str) -> CheckoutBuilder:
        self._weighed.append(_Weighed(product, Weight(kg)))
        return self

    def with_combo_deal(self, sku_a: str, sku_b: str, deal_cents: int) -> CheckoutBuilder:
        self._combo_deals.append(ComboDeal(sku_a, sku_b, Money(deal_cents)))
        return self

    def build(self) -> Checkout:
        checkout = Checkout(self._combo_deals)
        for s in self._scanned:
            for _ in range(s.count):
                checkout.scan(s.product)
        for w in self._weighed:
            checkout.scan_weighted(w.product, w.weight)
        return checkout

File: supermarket-pricing/typescript/tests/CheckoutBuilder.ts
import { Checkout } from '../src/Checkout.js';
import { ComboDeal } from '../src/ComboDeal.js';
import { Money } from '../src/Money.js';
import type { Product } from '../src/Product.js';
import { Weight } from '../src/Weight.js';

interface Scanned { product: Product; count: number; }
interface Weighed { product: Product; weight: Weight; }

export class CheckoutBuilder {
  private readonly _scanned: Scanned[] = [];
  private readonly _weighed: Weighed[] = [];
  private readonly _comboDeals: ComboDeal[] = [];

  withScanned(product: Product, count = 1): this {
    this._scanned.push({ product, count });
    return this;
  }

  withWeighed(product: Product, kg: number): this {
    this._weighed.push({ product, weight: new Weight(kg) });
    return this;
  }

  withComboDeal(skuA: string, skuB: string, dealCents: number): this {
    this._comboDeals.push(new ComboDeal(skuA, skuB, new Money(dealCents)));
    return this;
  }

  build(): Checkout {
    const checkout = new Checkout(this._comboDeals);
    for (const { product, count } of this._scanned) {
      for (let i = 0; i < count; i++) {
        checkout.scan(product);
      }
    }
    for (const { product, weight } of this._weighed) {
      checkout.scanWeighted(product, weight);
    }
    return checkout;
  }
}
